File: implementations/dotnet/pcf-dcp/src/Pcf.Dcp/DcpHeader.cs
using System;

namespace Pcf.Dcp;

/// <summary>The fixed 24-byte DCP Header at arena offset 0 (spec Section 6).</summary>
public sealed class DcpHeader
{
    /// <summary>PCF-DCP profile major version.</summary>
    public byte ProfileVersionMajor { get; set; }

    /// <summary>PCF-DCP profile minor version.</summary>
    public byte ProfileVersionMinor { get; set; }

    /// <summary>Reserved; MUST be 0 in v1.0.</summary>
    public ushort Flags { get; set; }

    /// <summary>Arena-relative offset of the first Inner Table Block (0 = none).</summary>
    public ulong InnerTableOffset { get; set; }

    /// <summary>Bump pointer: arena-relative offset of the first free byte.</summary>
    public ulong ArenaUsed { get; set; }

    /// <summary>Serialise to the on-disk 24-byte layout.</summary>
    public byte[] ToBytes()
    {
        var b = new byte[Constants.DcpHeaderSize];
        Buffer.BlockCopy(Constants.DcpMagic, 0, b, 0, 4);
        b[4] = ProfileVersionMajor;
        b[5] = ProfileVersionMinor;
        LittleEndian.WriteU16(b, 6, Flags);
        LittleEndian.WriteU64(b, 8, InnerTableOffset);
        LittleEndian.WriteU64(b, 16, ArenaUsed);
        return b;
    }

    /// <summary>Parse from the on-disk 24-byte layout, validating the magic.</summary>
    public static DcpHeader FromBytes(byte[] b)
    {
        for (int i = 0; i < 4; i++)
        {
            if (b[i] != Constants.DcpMagic[i])
            {
                throw PcfDcpException.BadDcpMagic();
            }
        }
        return new DcpHeader
        {
            ProfileVersionMajor = b[4],
            ProfileVersionMinor = b[5],
            Flags = LittleEndian.ReadU16(b, 6),
            InnerTableOffset = LittleEndian.ReadU64(b, 8),
            ArenaUsed = LittleEndian.ReadU64(b, 16),
        };
    }

    /// <summary>Read a DCP Header from the start of an arena byte array.</summary>
    public static DcpHeader Read(byte[] arena)
    {
        if (arena.Length < Constants.DcpHeaderSize)
        {
            throw PcfDcpException.BadDcpMagic();
        }
        var fixedBytes = new byte[Constants.DcpHeaderSize];
        Buffer.BlockCopy(arena, 0, fixedBytes, 0, Constants.DcpHeaderSize);
        return FromBytes(fixedBytes);
    }
}

File: implementations/php/pcf-dcp/src/DcpHeader.php
<?php

declare(strict_types=1);

namespace Kduma\PCFDCP;

/** The fixed 24-byte DCP Header at arena offset 0 (spec Section 6). */
final class DcpHeader
{
    public function __construct(
        public int $profileVersionMajor,
        public int $profileVersionMinor,
        public int $flags,
        public int $innerTableOffset,
        public int $arenaUsed,
    ) {
    }

    /** Serialise to the on-disk 24-byte layout. */
    public function toBytes(): string
    {
        return Consts::DCP_MAGIC
            . \chr($this->profileVersionMajor & 0xFF)
            . \chr($this->profileVersionMinor & 0xFF)
            . pack('v', $this->flags & 0xFFFF)
            . pack('P', $this->innerTableOffset)
            . pack('P', $this->arenaUsed);
    }

    /** Parse from the on-disk 24-byte layout, validating the magic. */
    public static function fromBytes(string $b): self
    {
        if (\strlen($b) < Consts::DCP_HEADER_SIZE || substr($b, 0, 4) !== Consts::DCP_MAGIC) {
            throw PcfDcpException::badDcpMagic();
        }

        return new self(
            \ord($b[4]),
            \ord($b[5]),
            unpack('v', substr($b, 6, 2))[1],
            unpack('P', substr($b, 8, 8))[1],
            unpack('P', substr($b, 16, 8))[1],
        );
    }

    /** Read a DCP Header from the start of an arena byte string. */
    public static function read(string $arena): self
    {
        if (\strlen($arena) < Consts::DCP_HEADER_SIZE) {
            throw PcfDcpException::badDcpMagic();
        }

        return self::fromBytes(substr($arena, 0, Consts::DCP_HEADER_SIZE));
    }
}
